text/0000-run-to-completion-async-fn.md

```diff
@@ -0,0 +1,908 @@
+
+Any code which is similiar to [`select!`](https://docs.rs/futures/0.3.4/futures/macro.select.html),
+and which was not converted to be made aware of `poll_drop`
+could cause `poll_drop` not be called. Therefore `poll_drop` is not a mechanism
+which e.g. allows us to add a safe Rust APIs around async completion based
+mechanism.
+
+`#[completion] async fn` in comparison enforces through the type system that
+methods will run to completion, and any necessary cleanup code is run before 
+resources are released.
+
+`poll_drop` and `#[completion] async fn` are also similar in the sense that they
+both require some `Future`s to be rewritten to support proper cleanup. The
+main difference here is that `#[completion] async fn` only requires new `Future`s
+that want to support running to completion to be rewritten, and will enforce
+this constraint through Rusts type system. `poll_drop` in comparison would require
+potentially every existing `Future` and `Future`-combinator to be changed, in
+order to provide a similary high guarantee of async cleanup code being performed.
+
+There exists also the possibility to introduce a combination of both proposals:
+
+Since `RunToCompletionFuture` is a new trait, it would be possible to add
+`fn poll_drop(self: Pin<&mut Self>)` on this trait, and enforce users to of the
+trait to use it: The `unsafe` contract in the trait could require callers to
+either `.poll()` the future to completion, or to switch over to calling `poll_drop`.
+
+However the benefit seems low. By doing this the Future would again move away
+from one linear code path into an alternate one - which is what run to completion
+wants to avoid. It is also highly likely that the code inside `poll_drop` would
+actually be equal to the one `poll` itself, since `poll` is also expected to
+perform any final cleanup.
+
+A different alternative might be to pair run to completion methods later with
+some kind of `finally` or `defer` blocks, which allow to run some (potentially
+async) code also if methods are exiting early. The benefit of this approach is
+that it could work for synchronous methods as well as for async methods.
+
+# Prior art
+[prior-art]: #prior-art
+
+Examples for languages which added support for async/await are:
+- C#
+- Javascript
+- C++ (Coroutines)
+- Kotlin (Coroutines)
+- Zig
+
+In all these languages async functions have run to completion semantics. Rust
+is the only language that is known to the author which allows to stop the
+execution of an asynchronous function in the middle of the execution. By adopting
+this RFC, Rust would gain a mechanism which optionally allows Rusts asynchronous
+functions closer to the ones found in other languages, as well as closer to
+synchronous functions.
+
+The `CancellationToken/StopToken` approach, which provides the ability to
+cooperatively cancel methods with run to completion semantics also had been
+successfully deployed in these environments.
+
+# Unresolved questions
```

Comment:
Curious about crash consistency, and how this works in the case of panics.

Reply:
If we implement `poll_drop`, then panics could run `poll_drop` to completion as the stack is being unwound.  C++ takes a similar approach via `unhandled_exception`.